File: dynamic_programming/subarray_and_subsequence/152_maximum_product_subarray.cc
#include <algorithm>
#include <vector>

// 2024/01/12: 直接看答案.

// 关键思路: 想到用两个dp数组同时存储最大乘积和最小乘积.
int MaxProduct(std::vector<int>& nums) {
  std::vector<int> max_f(nums), min_f(nums);
  for (int i = 1; i < nums.size(); ++i) {
    max_f[i] =
        std::max({max_f[i - 1] * nums[i], min_f[i - 1] * nums[i], nums[i]});
    min_f[i] =
        std::min({min_f[i - 1] * nums[i], max_f[i - 1] * nums[i], nums[i]});
  }
  return *std::max_element(max_f.begin(), max_f.end());
}

File: dynamic_programming/subarray_and_subsequence/152_maximum_product_subarray.py
from typing import List


# 2026/06/08: 直接看答案.
def max_product(nums: List[int]) -> int:
    min_dp, max_dp = list(nums), list(nums)

    for i in range(1, len(nums)):
        min_dp[i] = min(min_dp[i - 1] * nums[i], max_dp[i - 1] * nums[i], nums[i])
        max_dp[i] = max(min_dp[i - 1] * nums[i], max_dp[i - 1] * nums[i], nums[i])
    return max(max_dp)


nums = [2, 3, -2, 4]
print(max_product(nums))
